### ISSUE TITLE
Second call EmbeddedStorage.start(root, targetDirectory) with another root raise exception: Cannot cast ...

### ISSUE THREAD

**Issue by zdenek-jonas** [microstream-one/microstream#139]
one.microstream.exceptions.TypeCastException: Cannot cast one.microstream.collections.BulkList to test.microstream.handler.BinaryHandlerBulkListTest$BulkListData

I have started the storage with one type as root, then shutdown. 
Then clean i the storage:
`FileUtils.deleteDirectory(targetDirectory.toFile());`
Then i started this storage again with another root:
`EmbeddedStorage.start(root, targetDirectory);`

I get:
```
one.microstream.exceptions.TypeCastException: Cannot cast one.microstream.collections.BulkList to test.microstream.handler.BinaryHandlerBulkListTest$BulkListData

	at one.microstream.persistence.binary.internal.AbstractBinaryHandlerReflective.updateState(AbstractBinaryHandlerReflective.java:521)
	at one.microstream.persistence.binary.internal.AbstractBinaryHandlerReflective.updateState(AbstractBinaryHandlerReflective.java:1)
	at one.microstream.persistence.binary.types.BinaryLoader$Default.buildInstances(BinaryLoader.java:447)
	at one.microstream.persistence.binary.types.BinaryLoader$Default.build(BinaryLoader.java:382)
	at one.microstream.persistence.binary.types.BinaryLoader$Default.get(BinaryLoader.java:825)
	at one.microstream.persistence.binary.types.BinaryLoader$Default.loadRoots(BinaryLoader.java:869)
	at one.microstream.storage.types.EmbeddedStorageManager$Default.loadExistingRoots(EmbeddedStorageManager.java:304)
	at one.microstream.storage.types.EmbeddedStorageManager$Default.initialize(EmbeddedStorageManager.java:324)
	at one.microstream.storage.types.EmbeddedStorageManager$Default.start(EmbeddedStorageManager.java:212)
	at one.microstream.storage.types.EmbeddedStorageManager$Default.start(EmbeddedStorageManager.java:1)
	at one.microstream.storage.types.EmbeddedStorage.createAndStartStorageManager(EmbeddedStorage.java:660)
	at one.microstream.storage.types.EmbeddedStorage.start(EmbeddedStorage.java:518)
	at test.microstream.handler.AbstractHandlerTest.startStorage(AbstractHandlerTest.java:122)
	at test.microstream.handler.AbstractHandlerTest.saveAndReload(AbstractHandlerTest.java:113)
	at test.microstream.handler.BinaryHandlerBulkListTest.bulkListAlsRootTest(BinaryHandlerBulkListTest.java:25)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:675)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at 
```

Harald prepare now the simple main methods to reproduce this problem. He will add this code into this issue.

**Comment by hg-ms** [comment]
Here is a simple class that creates a similar problem if a new microstream instance created delete and created again:
```
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Comparator;

import one.microstream.storage.types.EmbeddedStorage;
import one.microstream.storage.types.EmbeddedStorageManager;

public class FailedSecondStart
{
	static final Path WORKING_DIR = Paths.get("storage");
	
	public static void main(final String[] args) throws IOException
	{
		startStorage();
		
		cleanup();
		
		startStorage();
	}
	
	private static void cleanup() throws IOException
	{
	      Files.walk(WORKING_DIR)
	      .sorted(Comparator.reverseOrder())
	      .map(Path::toFile)
	      .forEach(File::delete);
		
	}

	public static void startStorage()
	{
		final EmbeddedStorageManager storage = EmbeddedStorage.start(WORKING_DIR);
		final String myRoot = new String("Root");
		
		storage.store(myRoot);		
		
		final Object root = storage.root();
		System.out.println("stored string: " + root);
		
		storage.shutdown();
	}	
}
```

edit TM:
The code example has so many problems I need to compile them here:

- The constant is not a constant (missing "final" and wrong naming). I fixed that by editing but it makes no sense to fix every problem retrospectively.
- a method called "startStorage" actually performs the test and stops the storage in the end. This is misleading and weird. It should be called "runTest" or something like that.
- A String is instantiated by calling `new String("...")`. But `"..."` already is a String!
- But most importantly: this storage has no root instance associated. Just naming a local variable something with "root" in it does not make its referenced instance the root of the storage. All this code does is store a String with the content 'R', 'o', 'o', 't' that is never referenced by ANYTHING. See the numerous EmbeddedStorage#start variants for associating a root instance BEFORE starting the storage and EmbeddedStorageManager#setRoot for associating a root instance AFTER starting.

This really bugs me. I did several iterations of consolidating and extremely simplifying root handling for the user at the cost of some borderline insanely complex internal logic to cover all the different cases people might see as the "natural" way to use a root instance (before/after start etc.). And STILL someone finds a way to fail at even the most trivial way to define a root. The library can't just magically know which instance is supposed to be the root. You have to at least tell it what the root instance is supposed to be.
I don't know how to make it even easier to understand than it already is. I'm done. I got nothing.

**Comment by tm-ms** [comment]
On the actual problem:

It took some debugging, but I found the cause for the failed second run.

Here is what happens:
- There is a AFS singleton instance. More precisely: by default, it is a NioFileSystem instance.
- It acts as a kind of a secondary file system instance above the physical file system. 
- But from microstream's perspective, the AFS instance is more like the primary file system and the physical file system is the secondary. Because that was the whole point of the IO-layer abstraction.
- The first test run registers physical files at the AFS instance.
- Everything works fine.
- But the cleanup only deletes the files on the physical file system but NOT on in the AFS instance.
- The second test run uses the AFS instance, sees existing files there and uses them. But their physical equivalents are missing. Hence the error (failed validation for files with size 0, etc.)

I currently don't know how to avoid such problems systematically since this "schizophrenic" behavior was kind of the basic idea behind AFS.

I described the general problem in more detail in the separate issue microstream-one/microstream-private#363.

As far as this issue is concerned: Please check if the provided sample really reproduces the same problem as initially described as it seems to me this is a different one. But I might be wrong.

For the immediate problem, I added a check to the storage initialization to only consider AFS files that have existing physical equivalents. The code sample works now, but it does not solve the problem of residual inconsistent file entries on the AFS level.

**Comment by zdenek-jonas** [comment]
Ok, for these test i have create subfolder for each test. With this works. 
I will create another sample, with some real data.

**Comment by zdenek-jonas** [comment]
project: microstream-test
branch: bulklist
test location: /intern/src/test/java/Handler/BinaryHandlerBulkListTest

Attention please, branch is not a master for this test.

**Comment by zdenek-jonas** [comment]
is already fixed
